Fix the reverb effect

timrae opened the pull request:
The previous implementation of the reverb effect was suffering from serious artifacts in terms of feedback / distortion that the original LADSPA plugin doesn't suffer from:
https://bugs.launchpad.net/mixxx/+bug/1647568

I updated the code base to use the latest version (0.9.24) from the original author (http://quitte.de/dsp/caps.html), and in doing this I disabled the unused code with preprocessor directives instead of deleting it, in order to make it easier to update in the future.

Next I tried to track down where the bug was. It turns out that the `SampleUtil::linearCrossfadeBuffers()` call was causing all the feedback problems, and it seems to be safe to remove this.

This commit also adds the missing two controls for damping and blend, and remaps the meta knob from bandwidth to blend, which fixes the following bug:
https://bugs.launchpad.net/mixxx/+bug/1647775

timrae commented:
It may also fix this, though I haven't checked:
https://bugs.launchpad.net/mixxx/+bug/1647547

rryan commented on src/effects/native/reverbeffect.cpp:
```diff
@@ -74,48 +99,35 @@ void ReverbEffect::processChannel(const ChannelHandle& handle,
+                                 const CSAMPLE blendParam) {
+    effect.setBandwidth(bandwidth);
+    effect.setDamping(damping);
+    const auto decay = .749*decayParam;
```
Nit -- use float literals for float math (CSAMPLE is a float)

daschuer replied:
What is .749 here? Can we use a const variable to name it? 
IMHO we should not use auto in this case, because it is not easy to guess which type is decay.
A Bad example in our style guide

rryan commented on src/effects/native/reverbeffect.h:
```diff
@@ -16,24 +16,19 @@
 #include "util/sample.h"
 #include "util/types.h"
 
+const CSAMPLE MIN_VALUE = 0.0005;
```
Looks like these aren't used in the header. Could you put these in an anonymous namespace in reverbeffect.cpp and name them like 

```c++
namespace { 
const CSAMPLE kMinValue = ...;
}  // namespace
```

timrae replied:
Oops yes

rryan commented on src/effects/native/reverbeffect.cpp:
```diff
@@ -74,48 +99,35 @@ void ReverbEffect::processChannel(const ChannelHandle& handle,
+    DSP::FPTruncateMode _truncate;
+
+    for (auto i = 0u; i + 1 < numSamples; i += 2) {
+        CSAMPLE mono_sample = (in[i] + in[i + 1]) / 2;
```
Oops, ```/ 2.0f``` otherwise it converts to double and back. (the original code did this too)

timrae replied:
Nice catch

daschuer replied:
float / int is float! An it is perfectly clean and effective code, since the CPU has an instruction for it. 
Please keep the int 2.

rryan replied:
Huh -- I think I was pre-coffee, I swear I saw a "2." :)

rryan commented on src/effects/native/reverbeffect.cpp:
```diff
@@ -74,48 +99,35 @@ void ReverbEffect::processChannel(const ChannelHandle& handle,
+    effect.setBandwidth(bandwidth);
+    effect.setDamping(damping);
+    const auto decay = .749*decayParam;
+    const auto blend = pow(blendParam, 1.53);
```
hm, this is probably using ```pow(double, double)``` too

daschuer replied:
please define a constant for 1.53 as well.
auto is ambiguous here for the reader, since the type of pow is out of sight.

Be-ing commented:
Wow this sound so much better! Thank you!

I think it would be more useful to replace the Blend parameter with a Send parameter that works like the Send parameter for echo. That is, it should control the signal coming into the effect, not the output. When it is turned all the way down, the effect should continue the reverberations that were in progress until they decay out instead of cutting them off.

rryan commented on src/effects/native/reverbeffect.cpp:
```diff
@@ -27,36 +27,61 @@ EffectManifest ReverbEffect::getManifest() {
+    damping->setDefault(MIN_VALUE);
+    damping->setMaximum(MAX_VALUE);
+
+    EffectManifestParameter* blend = manifest.addParameter();
```
Hm, I wonder if effect slots should have a mix parameter (like the chain does). Seems like each effect shouldn't have to provide its own implementation of dry/wet.

Be-ing replied:
My concern with adding a mix parameter for every effect is cluttering the interface. Would it really be helpful? Clearly it is for reverb and echo, but I'm not sure about other effects.

Be-ing replied:
Maybe we need a send knob for the whole chain?

Be-ing replied:
Or maybe the dry/wet for the chain should be replaced with send (control input signal instead of output)?

timrae replied:
The design of the chain itself is as an "insert" effect, not as a "send" effect, so dry/wet makes more sense.  I'm doing it the exact same way as the original LADSPA plugin which makes it easier to verify the correctness, and I think it's common to have wet/dry on a reverb effect intended as an insert. For example see pantheon:

https://www.kvraudio.com/product/pantheon-by-lexicon

Most traditional insert effects like eq don't require a wet/dry knob so I don't think it should be required.

timrae replied:
If we were to change anything, I'd remove the wet / dry control on the effect unit altogether. It's unusual and not required now that we have meta knobs, as effects like delay/reverb which usually require wet dry parameters can map their local wet dry to the meta knob

rryan replied:
In case you haven't seen it -- Mixxx effect chains can swap between insert ```output = input * (1 - wet) + effect(input) * wet```  and send (```output = input  + effect(input)``` mode on the fly via the insertion_type control.

Be-ing replied:
I've been wondering what that insertion_type stuff was about. Would it make sense to define it at the effect/effect manifest level rather than chain?

rryan replied:
Speaking generally, the benefit of the effect system providing a feature over the effect itself doing it is that we get:

* Fewer bugs / better performance since we have one unit tested implementation that we can be sure gets vectorized on platforms we support. 
* Consistency of the interface (also fewer translations to do)
* Ability to provide a specific control so users can map it to their DJ controllers. (e.g. lots of controllers have dry/wet knobs and if D/W is an effect-defined parameter rather than something provided by the effect system then we can't easily provide a CO for it).

That's why I lean towards wanting the effect system to provide features rather than putting them into effect parameters. 

BTW -- it's likely I'm a reverb noob :), what do you use the blend parameter for that the chain D/W doesn't give you? Is it superknob linking?

rryan replied:
> I've been wondering what that insertion_type stuff was about. Would it make sense to define it at the effect/effect manifest level rather than chain?

Yea, that probably makes more sense. It'll come at a higher cost CPU-wise because right now we get away with some optimizations by assuming all the effects in a chain are either insert or send. 
https://github.com/mixxxdj/mixxx/blob/master/src/engine/effects/engineeffectchain.cpp#L180

To do it right, we'd have to do an insert or send mix for each effect, rather than just one for the chain.

Be-ing replied:
> what do you use the blend parameter for that the chain D/W doesn't give you?

Control over the individual effect. For all other effects, there is a way to set the metaknob so the effect is not heard. Without adding this control at the effect level for reverb, there is no way to turn the reverb off by turning the metaknob. The only way to turn it off is by turning the dry/wet all the way dry for the whole chain or toggling the enable button for the effect, which abruptly stops the effect.

timrae replied:
> That's why I lean towards wanting the effect system to provide features rather than putting them into effect parameters.

Yes, if all effects needed the feature then sure it makes sense. But if only some need it (as in this case), then it makes more sense to define them in the effect. In this case the code is ported directly from the original "manufacturer", so it's been tested in production for more than 10 years, and we shouldn't have anything to worry about.

> BTW -- it's likely I'm a reverb noob :), what do you use the blend parameter for that the chain D/W doesn't give you? Is it superknob linking?

Let's say you have a flanger, delay, and reverb all loaded into the chain. Naturally, you want to adjust the strength of each effect individually (e.g. via the meta parameter). With flanger we  use the depth, with delay we use the feedback, but with reverb there currently isn't anything. 

That's what blend is required for. The chain D/W changes the strength of all FX at once so it's not suitable. In fact I can't really think of a common use case for wanting the chain dry/wet... Most people are always going to leave it on 100% wet.

rryan replied:
> Yes, if all effects needed the feature then sure it makes sense. But if only some need it (as in this case), then it makes more sense to define them in the effect.

I think there are other alternatives for the "this doesn't make sense in context" problem -- like having the effect indicate to the effect system whether it makes sense to show a D/W knob or send knob. And a per-effect D/W or send could be optionally presented in the GUI along-side the regular parameters when appropriate. 

> In this case the code is ported directly from the original "manufacturer", so it's been tested in production for more than 10 years, and we shouldn't have anything to worry about.

The implementation of blend here is guaranteed to not be vectorized due to the function call in the loop body (see the loop body disassembled below). The dry/wet here is also not ramping, so discontinuities can produce artifacts if the blend parameter is changed too quickly. There are benefits to providing the support in the effects system.

```
mixxx[0x100026db0] <+288>: leal   -0x1(%r13), %r12d
mixxx`MixxxPlateX2::processBuffer + 292 at Reverb.cc:453
   452 		for (uint i = 0; i + 1 < frames; i += 2) {
   453 			sample_t mono_sample = (in[i] + in[i + 1]) / 2;
   454 			sample_t xl, xr;
mixxx[0x100026db4] <+292>: movl   %r13d, %r14d
mixxx[0x100026db7] <+295>: movss  (%rax,%r14,4), %xmm0      ; xmm0 = mem[0],zero,zero,zero
mixxx`MixxxPlateX2::processBuffer + 301 at Reverb.cc:453
   452 		for (uint i = 0; i + 1 < frames; i += 2) {
   453 			sample_t mono_sample = (in[i] + in[i + 1]) / 2;
   454 			sample_t xl, xr;
mixxx[0x100026dbd] <+301>: addss  (%rax,%r12,4), %xmm0
mixxx`MixxxPlateX2::processBuffer + 307 at Reverb.cc:453
   452 		for (uint i = 0; i + 1 < frames; i += 2) {
   453 			sample_t mono_sample = (in[i] + in[i + 1]) / 2;
   454 			sample_t xl, xr;
mixxx[0x100026dc3] <+307>: mulss  0x68b4f5(%rip), %xmm0     ; typeinfo name for ReplayGain + 68
mixxx`MixxxPlateX2::processBuffer + 315 at Reverb.cc:455
   454 			sample_t xl, xr;
   455 			PlateStub::process(mono_sample, decay, &xl, &xr);
   456 			out[i] = blend*xl + dry*in[i];
mixxx[0x100026dcb] <+315>: movq   %r15, %rdi
mixxx[0x100026dce] <+318>: movss  -0x44(%rbp), %xmm1        ; xmm1 = mem[0],zero,zero,zero
mixxx[0x100026dd3] <+323>: movq   %r15, %rbx
mixxx[0x100026dd6] <+326>: movq   %rax, %r15
mixxx[0x100026dd9] <+329>: leaq   -0x2c(%rbp), %rsi
mixxx[0x100026ddd] <+333>: leaq   -0x30(%rbp), %rdx
mixxx[0x100026de1] <+337>: callq  0x100026500               ; PlateStub::process at Reverb.cc:245
mixxx[0x100026de6] <+342>: movq   %r15, %rax
mixxx[0x100026de9] <+345>: movq   %rbx, %r15
mixxx[0x100026dec] <+348>: movss  -0x2c(%rbp), %xmm0        ; xmm0 = mem[0],zero,zero,zero
mixxx[0x100026df1] <+353>: movss  -0x48(%rbp), %xmm3        ; xmm3 = mem[0],zero,zero,zero
mixxx`MixxxPlateX2::processBuffer + 358 at Reverb.cc:456
   455 			PlateStub::process(mono_sample, decay, &xl, &xr);
   456 			out[i] = blend*xl + dry*in[i];
   457 			out[i + 1] = blend*xr + dry*in[i + 1];
mixxx[0x100026df6] <+358>: mulss  %xmm3, %xmm0
mixxx[0x100026dfa] <+362>: movss  (%rax,%r12,4), %xmm1      ; xmm1 = mem[0],zero,zero,zero
mixxx[0x100026e00] <+368>: movss  -0x3c(%rbp), %xmm2        ; xmm2 = mem[0],zero,zero,zero
mixxx`MixxxPlateX2::processBuffer + 373 at Reverb.cc:456
   455 			PlateStub::process(mono_sample, decay, &xl, &xr);
   456 			out[i] = blend*xl + dry*in[i];
   457 			out[i + 1] = blend*xr + dry*in[i + 1];
mixxx[0x100026e05] <+373>: mulss  %xmm2, %xmm1
mixxx`MixxxPlateX2::processBuffer + 377 at Reverb.cc:456
   455 			PlateStub::process(mono_sample, decay, &xl, &xr);
   456 			out[i] = blend*xl + dry*in[i];
   457 			out[i + 1] = blend*xr + dry*in[i + 1];
mixxx[0x100026e09] <+377>: addss  %xmm0, %xmm1
mixxx`MixxxPlateX2::processBuffer + 381 at Reverb.cc:456
   455 			PlateStub::process(mono_sample, decay, &xl, &xr);
   456 			out[i] = blend*xl + dry*in[i];
   457 			out[i + 1] = blend*xr + dry*in[i + 1];
mixxx[0x100026e0d] <+381>: movq   -0x38(%rbp), %rcx
mixxx[0x100026e11] <+385>: movss  %xmm1, (%rcx,%r12,4)
mixxx[0x100026e17] <+391>: movq   -0x38(%rbp), %rcx
mixxx[0x100026e1b] <+395>: movss  -0x30(%rbp), %xmm0        ; xmm0 = mem[0],zero,zero,zero
mixxx`MixxxPlateX2::processBuffer + 400 at Reverb.cc:457
   456 			out[i] = blend*xl + dry*in[i];
   457 			out[i + 1] = blend*xr + dry*in[i + 1];
   458 		}
mixxx[0x100026e20] <+400>: mulss  %xmm3, %xmm0
mixxx[0x100026e24] <+404>: movss  (%rax,%r14,4), %xmm1      ; xmm1 = mem[0],zero,zero,zero
mixxx`MixxxPlateX2::processBuffer + 410 at Reverb.cc:457
   456 			out[i] = blend*xl + dry*in[i];
   457 			out[i + 1] = blend*xr + dry*in[i + 1];
   458 		}
mixxx[0x100026e2a] <+410>: mulss  %xmm2, %xmm1
mixxx`MixxxPlateX2::processBuffer + 414 at Reverb.cc:457
   456 			out[i] = blend*xl + dry*in[i];
   457 			out[i + 1] = blend*xr + dry*in[i + 1];
   458 		}
mixxx[0x100026e2e] <+414>: addss  %xmm0, %xmm1
mixxx`MixxxPlateX2::processBuffer + 418 at Reverb.cc:457
   456 			out[i] = blend*xl + dry*in[i];
   457 			out[i + 1] = blend*xr + dry*in[i + 1];
   458 		}
mixxx[0x100026e32] <+418>: movss  %xmm1, (%rcx,%r14,4)
mixxx[0x100026e38] <+424>: movl   -0x40(%rbp), %ecx
mixxx`MixxxPlateX2::processBuffer + 427 at Reverb.cc:452
   451 		// loop through the buffer, processing each sample
   452 		for (uint i = 0; i + 1 < frames; i += 2) {
   453 			sample_t mono_sample = (in[i] + in[i + 1]) / 2;
mixxx[0x100026e3b] <+427>: addl   $0x2, %r13d
mixxx[0x100026e3f] <+431>: cmpl   %ecx, %r13d
mixxx[0x100026e42] <+434>: jb     0x100026db0               ; <+288> at Reverb.cc:453
```

rryan commented on src/effects/native/reverbeffect.cpp:
```diff
@@ -74,48 +99,35 @@ void ReverbEffect::processChannel(const ChannelHandle& handle,
 
-    for (uint i = 0; i + 1 < numSamples; i += 2) {
-        CSAMPLE mono_sample = (pInput[i] + pInput[i + 1]) / 2;
+void ReverbEffect::processBuffer(MixxxPlateX2& effect, const CSAMPLE* in, CSAMPLE* out, const unsigned int numSamples,
```
We follow Google style here with only using const reference arguments:
https://google.github.io/styleguide/cppguide.html#Reference_Arguments
Could you pass MixxxPlateX2 as a pointer instead?

timrae replied:
At my work we recently changed all the pointers that could be refs back to refs... The rule doesn't feel very modern, and cppcoreguideines for example doesn't have it. Won't you reconsider? 

It feels wrong to have to dereference a pointer in this case.

ywwg replied:
As a project we've decided to stick to the style guide so that the code is consistent and all parts of the code are readable for all of our developers.  While there are occasionally cases where the style guide can feel restricting or arbitrary, the benefits of having a consistent codebase, especially over very long development periods, makes it worth it to us.  While some old crufty parts of the code may not be conforming, we strongly prefer that new code follow the guide (with our slight modifications (https://www.mixxx.org/wiki/doku.php/coding_guidelines))

daschuer replied:
> It feels wrong to have to dereference a pointer in this case.

The issue of using non cons references is that you cannot distinguish them from copy by value parameter at the usage place. You nee to scroll up to check if this action stays in the function or effects the calling code. A const reference can always treated as a local copy. 

Using the -> operator is a clear sign that our action effects the surrounding code.

rryan commented on src/effects/native/reverbeffect.cpp:
```diff
@@ -27,36 +27,61 @@ EffectManifest ReverbEffect::getManifest() {
     damping->setName(QObject::tr("Damping"));
     damping->setDescription(QObject::tr("Higher damping values cause "
-            "reverberations to die out more quickly."));
+            "high frequencies to decay more quickly than low frequencies."));
```
Note, that changing a translation string requires re-translating this to all our supported languages. If there isn't a strong need to reword, I would leave as it was.

Be-ing replied:
The meaning has changed substantially, so I think there is a need.

rryan replied:
The meaning of damping changed?

rryan replied:
Oh wow, I missed that. NEvermind

rryan commented on lib/reverb/Reverb.cc:
```diff
@@ -45,16 +42,151 @@
 PlateStub::init()
 {
 	f_lfo = -1;
-	// TODO(owilliams): use actual sample rate.
```
Yikes, thanks for fixing this!

rryan commented:
Haven't tried it yet but thanks for updating us!

ywwg commented:
my apologies for making such a mess of it to begin with, thanks! :)

rryan commented on lib/reverb/basics.h:
```diff
@@ -104,11 +84,7 @@ T clamp (T value, T lower, T upper)
-{
-	return (float) rand() / (float) RAND_MAX;
-}
+static inline float frandom() { return (float) random() / (float) RAND_MAX; }
```
I think @ywwg probably changed this to ```rand``` since Windows doesn't have ```random```:

```
[CXX] lib\reverb\Reverb.cc
Reverb.cc
c:\projects\mixxx\lib\reverb\basics.h(36) : warning C4005: '_USE_MATH_DEFINES' : macro redefinition
        command-line arguments :  see previous definition of '_USE_MATH_DEFINES'
c:\projects\mixxx\lib\reverb\basics.h(87) : error C3861: 'random': identifier not found
scons: *** [lib\reverb\Reverb.obj] Error 2
scons: building terminated because of errors.
```

timrae commented:
I force pushed a new commit, moving all the looping code and scaling back to the library files, so now it's consistent with the rest of the code there, and we don't have any "magic numbers" in the mixxx part of the code.

rryan commented on src/effects/native/reverbeffect.cpp:
```diff
@@ -74,48 +99,15 @@ void ReverbEffect::processChannel(const ChannelHandle& handle,
 
-    bool params_changed = (damping != pState->prev_damping ||
-                           bandwidth != pState->prev_bandwidth);
+    // Set the sample rate (TODO XXX: Maybe we want to call .init() after this?)
```
We do need to call ```init()``` otherwise it has no effect (looks like fs is only used in ```init()```).

Maybe you could call it in setSampleRate if fs != sampleRate, or by adding a getSampleRate method and calling init from here if it doesn't match sampleRate.

rryan commented on lib/reverb/Reverb.cc:
```diff
@@ -458,3 +458,9 @@ void MixxxPlateX2::processBuffer(const sample_t* in, sample_t* out, const uint f
+void MixxxPlateX2::setSampleRate(const double sampleRate) {
+	if (fs != sampleRate) {
+		fs = sampleRate;
+		init();
```
I think we also need to call activate() since it sets some LFO state based on fs. 

This is only ever going to get called the first time an effect is processed after creation, or when the user changes the sample rate on their soundcard (so the sound is going to drop out anyway).

timrae replied:
Done

timrae commented:
In terms of the issue of whether to modify the FX model to include wet/dry at the effect level instead of the chain level (that discussion is buried in out of date commit messages), I think this is obviously outside the scope of this PR. But if we were to explore this possibility, I think it would be convenient if the insert/send property were indeed defined at the effect level, and in only the latter case would a "send amount" knob be automatically shown in the skin. In that case we can easily just remove that parameter from `getManifest()` and send `1` for the W/D parameter to the reverb library.

Anyway, since that is out of scope, please go ahead and merge if everything looks good ;)

Be-ing commented:
I still do think the new knob should be a send knob that works like the send knob on the echo effect. If you turn that all the way down, the effect sound doesn't abruptly end; it keeps echoing out the signal the came in before the send was cut off.

timrae commented:
> I still do think the new knob should be a send knob that works like the send knob on the echo effect. If you turn that all the way down, the effect sound doesn't abruptly end; it keeps echoing out the signal the came in before the send was cut off.

True, that would be a real advantage, it would lead to much smoother transitions. However, I think it's outside the scope of this PR which is intended to update the underlying effect to the latest version, fix the feedback issue, and expose all of the controls (wet/dry is one of these). Finally, it's still not clear whether the change you are requesting should be done at the underlying plugin level, the mixxx effect level (src/effect/native), or at the mixxx effect framework level.

daschuer commented:
The Chain D/W knob an the insert type, is IMHO somehow a legacy concept now. 

One original Idea, was like implemented in LateNight that you use whole chains during the mix as you use effects in other softwares and not single effects. The single effects could be treat as more atomic stages like delay, filter which are combined a fancy effect like Reverb. From this point of view a chain D/W makes perfectly sense. 

The recent changes with the meta knob and the controller focus, manifest the alternative concept where the user uses the single effects during the mix. Now the Super knob is "only" a fancy gimmick and the D/W knob is IMHO broken. Since it works only as desired if the effect chain does not introduce a face shift. If It is used on effects with phase shift the D/W knob acts like a fancy filter, which does D/W but also introduces a heavy distortion around 50 % 

Only the effect itself can decide how to implement a D/W feature right. 
You can try this using an EQ. A full low kill setup with 50 % dry sounds differently than a 50 % kill, which should IMHO not the case.

timrae commented:
> The Chain D/W knob an the insert type, is IMHO somehow a legacy concept now.

Yes I completely agree, I vote to remove it from the skins and deprecate its control

> Only the effect itself can decide how to implement a D/W feature right.

Also completely agree, for the case of insert effects. If we could set insert/send behavior at the individual effect level, and if an effect declared itself to be of the SEND type, then I think it would be both safe and useful for us to handle the send amount control automatically.

timrae commented:
One of the MSVC tests failed (seems to be unrelated to the code); could a project owner please restart that build?

Be-ing commented:
> However, I think it's outside the scope of this PR which is intended to update the underlying effect to the latest version, fix the feedback issue, and expose all of the controls (wet/dry is just one of these). Finally, it's still not clear whether the change you are requesting should be done at the underlying plugin level, the mixxx effect level (src/effect/native), or at the mixxx effect framework level.

Whether it's done in this PR or a follow up soon after this is a little procedural detail that I don't really care about, but I do think it should be taken care of for the 2.1 release. I think the effect's dry/wet parameter should be hard coded at fully wet and the Mixxx effect code around it can implement the send parameter. We can keep discussing potential changes to the effect framework, but I think broader changes to the effect framework should wait for after 2.1. Until then, implementing a send knob for this particular effect is a decent way to be able to have smooth fade outs with the reverb effect for 2.1.

timrae commented:
> Whether it's done in this PR or a follow up soon after this is a little procedural detail that I don't really care about

Not really... In the former there's a requirement for me to implement it myself, where as in the latter there isn't.

Be-ing commented:
> One original Idea, was like implemented in LateNight that you use whole chains during the mix as you use effects in other softwares and not single effects... The recent changes with the meta knob and the controller focus, manifest the alternative concept where the user uses the single effects during the mix.

These are not mutually exclusive and an effect unit may be used each way at different times.

Be-ing commented:
Btw, was the author of the effect ever contacted to ask if they'd be willing to provide a GPL exception for the Mac App Store?

rryan commented:
> Whether it's done in this PR or a follow up soon after this is a little procedural detail that I don't really care about, but I do think it should be taken care of for the 2.1 release.

One thing to be aware of is that if #1092 is merged for 2.1, we will start to get into the business of versioning of effects and needing to migrate user's effects configs as we change around what parameters each effect has. :-/ Up till now we've been free to change parameters without having to worry about it.

> Not really... In the former there's a requirement for me to implement it myself, where as in the latter there isn't.

Right, though you are adding new parameters so it is fair to question whether it would be better addressed at the system level rather than in the effect. I think discussion about this is in the scope of this PR -- but if it turns out it's better addressed at the system level then we would simply drop the parameter from this PR, not gate the PR on the bigger changes.

rryan commented:
> Btw, was the author of the effect ever contacted to ask if they'd be willing to provide a GPL exception for the Mac App Store?

Yes, they did not agree :(.

rryan commented:
> One of the MSVC tests failed (seems to be unrelated to the code); could a project owner please restart that build?

Done (it's probably fine if the x86 build passed)

rryan commented:
Huh, noticed something weird while looking at the disassembly for PlateStub::process, it looks like every 5th instruction or so is a double -> float or float -> double conversion. There are 72 conversions in the function with an -O3 build using clang.  :-/.  

http://paste.debian.net/908052/

Be-ing commented:
> if it turns out it's better addressed at the system level then we would simply drop the parameter from this PR, not gate the PR on the bigger changes.

It may be better addressed at the system level, but should we be making system level changes when we're trying to get a release out?

timrae commented:
> Huh, noticed something weird while looking at the disassembly for PlateStub::process, it looks like every 5th instruction or so is a double -> float or float -> double conversion. There are 72 conversions in the function with an -O3 build using clang. :-/.

It's not that surprising... I believe it's usual in audio DSP to convert from float to double during processing / arithmetic, and then back again to float in order to prevent compounding of rounding errors.

timrae commented:
@rryan [wrote](https://github.com/mixxxdj/mixxx/pull/1113#discussion_r95320200):
> The implementation of blend here is guaranteed to not be vectorized due to the function call in the loop body (see the loop body disassembled below). The dry/wet here is also not ramping, so discontinuities can produce artifacts if the blend parameter is changed too quickly. There are benefits to providing the support in the effects system.

I completely agree that adding support at the system level for either an optional wet/dry or send amount knob would be beneficial. It's not something that I'm personally interested in working on right now, however.

@Be-ing 
It seems like we agree that the effect might be better suited to use send amount instead of wet/dry. If we were to do that in this PR (with the possibility of taking it out when alternatives exist at the system level), I guess it should look something like this (in `MixxxPlateX2::processBuffer()`):

```c++
	for (uint i = 0; i + 1 < frames; i += 2) {
		sample_t mono_sample = blend*(in[i] + in[i + 1]) / 2;
		sample_t xl, xr;
		PlateStub::process(mono_sample, decay, &xl, &xr);
		out[i] = xl + dry*in[i];
		out[i + 1] = xr + dry*in[i + 1];
	}
```

Can you check how it sounds? I don't have time right now.

timrae commented:
I've pushed a commit with the scheme I suggested above. It is definitely smoother transitioning than before, especially if you use a larger decay value (obviously). With this someone can add support for send/mix knobs at the system level later on, or improve the current scheme by introducing ramping on the parameter changes without functionally changing the user interface.

Be-ing commented:
It now segfaults when the effect is enabled. gdb backtrace:
```
Thread 38 "mixxx" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffec37fe700 (LWP 6253)]
0x0000000000690918 in ReverbEffect::processChannel (groupFeatures=..., enableState=EffectProcessor::ENABLING, sampleRate=44100, 
    numSamples=2048, pOutput=0x7fff4cc36010, pInput=0x7fff3bca5010, pState=<optimized out>, handle=..., this=<optimized out>)
    at src/effects/native/reverbeffect.cpp:107
107         pState->reverb.setSampleRate(sampleRate);
Missing separate debuginfos, use: dnf debuginfo-install alsa-plugins-pulseaudio-1.1.1-1.fc24.x86_64 dbus-libs-1.11.6-1.fc24.x86_64 jasper-libs-1.900.13-1.fc24.x86_64 kde-runtime-libs-16.08.2-1.fc24.x86_64 kdelibs-4.14.25-1.fc24.x86_64 krb5-libs-1.14.4-4.fc24.x86_64 libXfixes-5.0.3-1.fc24.x86_64 libXi-1.7.8-2.fc24.x86_64 libXrandr-1.5.1-1.fc24.x86_64 libXrender-0.9.10-1.fc24.x86_64 libXtst-1.2.3-1.fc24.x86_64 libdrm-2.4.73-1.fc24.x86_64 libgcc-6.2.1-2.fc24.x86_64 libicu-56.1-5.fc24.x86_64 libjpeg-turbo-1.5.0-4.fc24.x86_64 libpng-1.6.26-1.fc24.x86_64 libsndfile-1.0.27-1.fc24.x86_64 libstdc++-6.2.1-2.fc24.x86_64 libuuid-2.28.2-1.fc24.x86_64 libwebp-0.5.1-2.fc24.x86_64 mesa-dri-drivers-12.0.3-2.fc24.x86_64 mesa-libGL-12.0.3-2.fc24.x86_64 mesa-libglapi-12.0.3-2.fc24.x86_64 openssl-libs-1.0.2j-1.fc24.x86_64 orc-0.4.26-1.fc24.x86_64 pcre-8.39-6.fc24.x86_64 qt4-style-oxygen-5.7.5-1.fc24.x86_64 qtscriptbindings-0.2.0-11.fc24.x86_64 systemd-libs-229-16.fc24.x86_64
(gdb) bt
#0  0x0000000000690918 in PerChannelEffectProcessor<ReverbGroupState>::process(ChannelHandle const&, float const*, float*, unsigned int, unsigned int, EffectProcessor::EnableState, GroupFeatureState const&) (groupFeatures=..., enableState=EffectProcessor::ENABLING, sampleRate=44100, numSamples=2048, pOutput=0x7fff4cc36010, pInput=0x7fff3bca5010, pState=<optimized out>, handle=..., this=<optimized out>) at src/effects/native/reverbeffect.cpp:107
#1  0x0000000000690918 in PerChannelEffectProcessor<ReverbGroupState>::process(ChannelHandle const&, float const*, float*, unsigned int, unsigned int, EffectProcessor::EnableState, GroupFeatureState const&) (this=<optimized out>, handle=..., pInput=0x7fff3bca5010, pOutput=0x7fff4cc36010, numSamples=2048, sampleRate=44100, enableState=EffectProcessor::ENABLING, groupFeatures=...)
    at src/effects/effectprocessor.h:81
#2  0x00000000007cd331 in EngineEffect::process(ChannelHandle const&, float const*, float*, unsigned int, unsigned int, EffectProcessor::EnableState, GroupFeatureState const&) (this=0x17f1290, handle=..., pInput=0x7fff3bca5010, pOutput=0x7fff4cc36010, numSamples=numSamples@entry=2048, sampleRate=sampleRate@entry=44100, enableState=EffectProcessor::ENABLING, groupFeatures=...)
    at src/engine/effects/engineeffect.cpp:104
#3  0x00000000007d0b7b in EngineEffectChain::process(ChannelHandle const&, float*, unsigned int, unsigned int, GroupFeatureState const&) (this=0x17edaf0, handle=..., pInOut=pInOut@entry=0x7fff3bca5010, numSamples=numSamples@entry=2048, sampleRate=sampleRate@entry=44100, groupFeatures=...) at src/engine/effects/engineeffectchain.cpp:209
#4  0x00000000007d226c in EngineEffectRack::process(ChannelHandle const&, float*, unsigned int, unsigned int, GroupFeatureState const&) (this=<optimized out>, handle=..., pInOut=pInOut@entry=0x7fff3bca5010, numSamples=numSamples@entry=2048, sampleRate=sampleRate@entry=44100, groupFeatures=...) at src/engine/effects/engineeffectrack.cpp:51
#5  0x00000000007d30b7 in EngineEffectsManager::process(ChannelHandle const&, float*, unsigned int, unsigned int, GroupFeatureState const&) (this=<optimized out>, handle=..., pInOut=0x7fff3bca5010, numSamples=2048, sampleRate=44100, groupFeatures=...)
    at src/engine/effects/engineeffectsmanager.cpp:139
#6  0x00000000007e08a0 in EngineDeck::process(float*, int) (this=0x1f35560, pOut=0x7fff3bca5010, iBufferSize=2048)
    at src/engine/enginedeck.cpp:112
#7  0x00000000007ee6bd in EngineMaster::processChannels(int) (this=this@entry=0x163d580, iBufferSize=iBufferSize@entry=2048)
    at src/engine/enginemaster.cpp:318
#8  0x00000000007f6ebb in EngineMaster::process(int) (this=0x163d580, iBufferSize=iBufferSize@entry=2048)
    at src/engine/enginemaster.cpp:350
#9  0x0000000000b694ab in SoundManager::onDeviceOutputCallback(unsigned int) (this=<optimized out>, iFramesPerBuffer=iFramesPerBuffer@entry=1024) at src/soundio/soundmanager.cpp:576
#10 0x0000000000b65a84 in SoundDevicePortAudio::callbackProcessClkRef(unsigned int, float*, float const*, PaStreamCallbackTimeInfo const*, unsigned long) (this=0x1e2e310, framesPerBuffer=1024, out=0x42efac30, in=<optimized out>, timeInfo=<optimized out>, statusFlags=<optimized out>) at src/soundio/sounddeviceportaudio.cpp:947
#11 0x00007ffff7872906 in NonAdaptingProcess (bp=bp@entry=0x42e850f0, streamCallbackResult=streamCallbackResult@entry=0x7ffec37fdd28, hostInputChannels=0x0, hostOutputChannels=0x42e8c520, framesToProcess=<optimized out>) at src/common/pa_process.c:871
#12 0x00007ffff7874914 in PaUtil_EndBufferProcessing (bp=bp@entry=0x42e850f0, streamCallbackResult=streamCallbackResult@entry=0x7ffec37fdd28) at src/common/pa_process.c:1583
#13 0x00007ffff787d705 in CallbackThreadFunc (userData=0x42e85080) at src/hostapi/alsa/pa_linux_alsa.c:4328
#14 0x00007ffff23a75ca in start_thread () at /usr/lib64/libpthread.so.0
---Type <return> to continue, or q <return> to quit---
#15 0x00007fffefd470ed in clone () at /usr/lib64/libc.so.6
```

Be-ing commented on lib/reverb/Reverb.cc:
```diff
@@ -449,12 +449,13 @@ void MixxxPlateX2::processBuffer(const sample_t* in, sample_t* out, const uint f
 	DSP::FPTruncateMode _truncate;
 
 	// loop through the buffer, processing each sample
+	// note (timrae):treat the effect as SEND type instead of INSERT type for smoother parameter changes
```
It may be helpful to expand on this comment and note that the effects system might obsolete this in the future.

Be-ing commented on src/effects/native/reverbeffect.cpp:
```diff
@@ -23,6 +23,19 @@ EffectManifest ReverbEffect::getManifest() {
+    EffectManifestParameter* send = manifest.addParameter();
+    send->setId("send_amount");
+    send->setName(QObject::tr("Send"));
+    send->setDescription(QObject::tr("How much of the signal to send to the effect unit"));
```
"How much of the input signal to send to the effect" (not effect unit)

Be-ing commented:
Regarding ramping, the echo effect has [a ramping implementation](https://github.com/mixxxdj/mixxx/blob/master/src/effects/native/echoeffect.cpp#L138) for its send parameter. Perhaps that would be appropriate for this as well?

rryan commented:
> It's not that surprising... I believe it's usual in audio DSP to convert from float to double during processing / arithmetic, and then back again to float in order to prevent compounding of rounding errors.

If the precision is needed then you'd typically convert to double, do your work, then convert back at the end. This code is converting back and forth on ~every individual step of ```process()``` (and process is per-sample! so that's 71 * samples_per_buffer conversions per invocation of the reverb). Beyond losing the precision it went to double for in the first place, a conversion takes 2-3x longer in micro-ops and latency than a vectorized multiplication -- so this is definitely sub-optimal.

Whether or not this needs double-precision math is worth investigating, but I think this code wasn't written with the idea that sample_t was float in mind (yet when used in LV2 it's defined as LADSPA_Data which is float...).

rryan commented:
> It now segfaults when the effect is enabled. gdb backtrace:

Huh -- I can't see from this what's going wrong, but looking at init, maybe it wasn't meant to be called twice. We may need to re-create the MixxxPlateX2 in the ReverbGroupState instead.

Be-ing commented:
I am torn about parameter ordering. On one hand, it kinda makes sense to have the send parameter first. On the other hand, it is not very useful to have it first if it can be controlled with the metaknob. It may be more useful to have the parameter(s) controlled by the metaknob last, so when controllers are manipulating the individual parameters, they can adjust the first three parameters and switch back to controlling the metaknob to control the last one. If the knob with the default metaknob linking is in the first three, it is rendundant to be able to control it when adjusting individual parameters.

This applies to all effects, not just this one. Ideally in the future, [parameters will be arbitrarily rearrangable](https://bugs.launchpad.net/mixxx/+bug/1653325), so we won't have to decide for users.

rryan commented:
With this tweak: https://github.com/rryan/mixxx/commit/87ce35b03d6837923fc026d40c675e749acd4109
there are no more float<->double conversions in PlateStub::process. I wrote a benchmark for various buffer sizes -- looks like a 30% or so speedup:

This branch:
```
Benchmark        Time(ns)    CPU(ns) Iterations
-----------------------------------------------
BM_Reverb/32         1969       1968     499554
BM_Reverb/64         3826       3826     190457
BM_Reverb/128        7886       7884     136578
BM_Reverb/256       11092      11090      60614
BM_Reverb/512       24149      24143      32763
BM_Reverb/1024      46467      46464      15198
BM_Reverb/2k        98071      98034       7327
BM_Reverb/4k       184035     183967       3717
```

With https://github.com/rryan/mixxx/commit/87ce35b03d6837923fc026d40c675e749acd4109
```
Benchmark        Time(ns)    CPU(ns) Iterations
-----------------------------------------------
BM_Reverb/32         1151       1151     583343
BM_Reverb/64         2095       2095     328259
BM_Reverb/128        4202       4202     166158
BM_Reverb/256        7446       7446      95395
BM_Reverb/512       15118      15119      45276
BM_Reverb/1024      30266      30266      23249
BM_Reverb/2k        59913      59913      11147
BM_Reverb/4k       120742     120742       5525
```

timrae commented:
I guess the idea of the author is to use interpolation for the float to
double conversions... In which case changing the interpolate function to
take a double as input might be pointless.

On 11 Jan. 2017 04:44, "RJ Skerry-Ryan" <notifications@github.com> wrote:

> With this tweak: rryan@87ce35b
> <https://github.com/rryan/mixxx/commit/87ce35b03d6837923fc026d40c675e749acd4109>
> there are no more float<->double conversions in PlateStub::process. I
> wrote a benchmark for various buffer sizes -- looks like a 30% or so
> speedup:
>
> This branch:
>
> Benchmark        Time(ns)    CPU(ns) Iterations
> -----------------------------------------------
> BM_Reverb/32         1969       1968     499554
> BM_Reverb/64         3826       3826     190457
> BM_Reverb/128        7886       7884     136578
> BM_Reverb/256       11092      11090      60614
> BM_Reverb/512       24149      24143      32763
> BM_Reverb/1024      46467      46464      15198
> BM_Reverb/2k        98071      98034       7327
> BM_Reverb/4k       184035     183967       3717
>
> With rryan@87ce35b
> <https://github.com/rryan/mixxx/commit/87ce35b03d6837923fc026d40c675e749acd4109>
>
> Benchmark        Time(ns)    CPU(ns) Iterations
> -----------------------------------------------
> BM_Reverb/32         1151       1151     583343
> BM_Reverb/64         2095       2095     328259
> BM_Reverb/128        4202       4202     166158
> BM_Reverb/256        7446       7446      95395
> BM_Reverb/512       15118      15119      45276
> BM_Reverb/1024      30266      30266      23249
> BM_Reverb/2k        59913      59913      11147
> BM_Reverb/4k       120742     120742       5525
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/1113#issuecomment-271677462>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACsA4udFI7oq8fT-W717ZYKeVBL2az7_ks5rQ9-pgaJpZM4LedKp>
> .
>

rryan commented:
> I guess the idea of the author is to use interpolation for the float to
> double conversions... In which case changing the interpolate function to
> take a double as input might be pointless.

Are you talking about ```Delay::get_linear``` / ```Delay::get_cubic```?  They're interpolating with ```(x-floor(x))``` to get the fractional part, then applying an interpolation function (linear or cubic) between nearby samples. I don't think changing the argument to double there matters?

rryan commented on src/effects/native/reverbeffect.cpp:
```diff
@@ -77,7 +77,7 @@ EffectManifest ReverbEffect::getManifest() {
 ReverbEffect::ReverbEffect(EngineEffect* pEffect,
                              const EffectManifest& manifest)
-        : m_pSendParameter(pEffect->getParameterById("send")),
+        : m_pSendParameter(pEffect->getParameterById("send_amount")),
```
Oh was the segfault just a null pointer? I was scratching my head about why calling init() twice could segfault -- the worst I can see that happens is we leak some memory in Delay if calling init twice.

Be-ing commented on src/effects/native/reverbeffect.cpp:
```diff
@@ -99,11 +99,21 @@ void ReverbEffect::processChannel(const ChannelHandle& handle,
-    pState->reverb.setSampleRate(sampleRate);
+    if (!pState->didInit) {
+        // init() function for reverb needs sampleRate
+        // this doesn't allow changing the sampleRate but it doesn't strongly affect the sound
```
I tested and it does unexpectedly shift the pitch of the reverberations when changing the sample rate until Mixxx is restarted. That's not a terrible bug, but it should be avoided if possible.

Be-ing commented on lib/reverb/Reverb.h:
```diff
@@ -223,10 +223,14 @@ class PlateX2
 /// (timrae) Define our own interface instead of using the original LADAP plugin interface
```
typo: LADSPA

Be-ing reviewed (commented):
I think we should figure out why calling init and activate twice made it segfault before instead of hacking around that.

Be-ing commented:
> I am torn about parameter ordering.

I do think it would be better to put the send parameter at the end and link it to the metaknob by default. I can change that together with doing the same for echo in another PR after this is merged.

rryan commented:
> I think we should figure out why calling init and activate twice made it segfault before instead of hacking around that.

Unless I'm mistaken, the segfault was m_pSendParameter being null because the parameter ID used to look it up didn't match the manifest. If so calling init twice should be harmless other than some memory leaking.

timrae commented:
> the segfault was m_pSendParameter being null 

Ah yes you're right, I thought those were two separate segfaults; I incorrectly thought the segfault was reported before I changed to send.

I've fixed that, updated the order of the FX, fixed the typo, changed the FX version number to 1.1, squashed everything down to one commit, and updated the commit message. Please test again.

rryan commented:
> changed the FX version number to 1.1,

Could you keep it at 1? We only have to start incrementing once we start serializing effect info to user profiles.

timrae commented:
>> I guess the idea of the author is to use interpolation for the float to double conversions... In which case changing the interpolate function to take a double as input might be pointless.

> Are you talking about Delay::get_linear / Delay::get_cubic? They're interpolating with (x-floor(x)) to get the fractional part, then applying an interpolation function (linear or cubic) between nearby samples. I don't think changing the argument to double there matters?

After merging this could you make a new PR and ping me? It might be worth contacting the author... He might merge the changes upstream, or point out some issue we didn't understand.

rryan commented:
> After merging this could you make a new PR and ping me? It might be worth contacting the author... He might merge the changes upstream, or point out some issue we didn't understand.

Sure, will do.

timrae commented:
> Regarding ramping, the echo effect has a ramping implementation for its send parameter. Perhaps that would be appropriate for this as well?

I'm not sure, I saw that but it didn't look trivial so I skipped it. Feel free to make a new PR if this is something you're interested in working on ;)

Be-ing commented:
Now it's segfaulting again :/
```Thread 38 "mixxx" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffec37fe700 (LWP 26521)]
0x00000000004bb8b2 in Lattice::process (d=0, x=<optimized out>, this=0x41801310) at lib/reverb/Reverb.h:66
66                                      x -= d*y;
Missing separate debuginfos, use: dnf debuginfo-install mesa-dri-drivers-12.0.3-2.fc24.x86_64
(gdb) bt
#0  0x00000000004bb8b2 in MixxxPlateX2::processBuffer(float const*, float*, unsigned int, float, float, float, float) (d=0, x=<optimized out>, this=0x41801310) at lib/reverb/Reverb.h:66
#1  0x00000000004bb8b2 in MixxxPlateX2::processBuffer(float const*, float*, unsigned int, float, float, float, float) (_xr=<synthetic pointer>, _xl=<synthetic pointer>, decay=0.374500006, x=<optimized out>, this=0x418012e8) at lib/reverb/Reverb.cc:249
#2  0x00000000004bb8b2 in MixxxPlateX2::processBuffer(float const*, float*, unsigned int, float, float, float, float) (this=this@entry=0x418012e8, in=in@entry=0x7fff482f2010, out=out@entry=0x7fff482f2010, frames=frames@entry=2048, bandwidthParam=<optimized out>, decayParam=<optimized out>, dampingParam=<optimized out>, blendParam=<optimized out>) at lib/reverb/Reverb.cc:458
#3  0x0000000000689315 in PerChannelEffectProcessor<ReverbGroupState>::process(ChannelHandle const&, float const*, float*, unsigned int, unsigned int, EffectProcessor::EnableState, GroupFeatureState const&) (groupFeatures=..., enableState=EffectProcessor::ENABLING, sampleRate=<optimized out>, numSamples=2048, pOutput=0x7fff482f2010, pInput=0x7fff482f2010, pState=0x418012e0, handle=..., this=<optimized out>) at src/effects/native/reverbeffect.cpp:116
#4  0x0000000000689315 in PerChannelEffectProcessor<ReverbGroupState>::process(ChannelHandle const&, float const*, float*, unsigned int, unsigned int, EffectProcessor::EnableState, GroupFeatureState const&) (this=<optimized out>, handle=..., pInput=0x7fff482f2010, pOutput=0x7fff482f2010, numSamples=2048, sampleRate=<optimized out>, enableState=EffectProcessor::ENABLING, groupFeatures=...)
    at src/effects/effectprocessor.h:81
#5  0x00000000007c4fdb in EngineEffect::process(ChannelHandle const&, float const*, float*, unsigned int, unsigned int, EffectProcessor::EnableState, GroupFeatureState const&) (this=0x415ec320, handle=..., pInput=pInput@entry=0x7fff482f2010, pOutput=pOutput@entry=0x7fff482f2010, numSamples=numSamples@entry=2048, sampleRate=sampleRate@entry=44100, enableState=EffectProcessor::ENABLED, groupFeatures=...) at src/engine/effects/engineeffect.cpp:104
#6  0x00000000007c7f65 in EngineEffectChain::process(ChannelHandle const&, float*, unsigned int, unsigned int, GroupFeatureState const&) (this=0x17d2180, handle=..., pInOut=pInOut@entry=0x7fff482f2010, numSamples=numSamples@entry=2048, sampleRate=sampleRate@entry=44100, groupFeatures=...) at src/engine/effects/engineeffectchain.cpp:190
#7  0x00000000007ca24c in EngineEffectRack::process(ChannelHandle const&, float*, unsigned int, unsigned int, GroupFeatureState const&) (this=<optimized out>, handle=..., pInOut=pInOut@entry=0x7fff482f2010, numSamples=numSamples@entry=2048, sampleRate=sampleRate@entry=44100, groupFeatures=...) at src/engine/effects/engineeffectrack.cpp:51
#8  0x00000000007cb097 in EngineEffectsManager::process(ChannelHandle const&, float*, unsigned int, unsigned int, GroupFeatureState const&) (this=<optimized out>, handle=..., pInOut=0x7fff482f2010, numSamples=2048, sampleRate=44100, groupFeatures=...)
    at src/engine/effects/engineeffectsmanager.cpp:139
#9  0x00000000007d8880 in EngineDeck::process(float*, int) (this=0x1e4f1c0, pOut=0x7fff482f2010, iBufferSize=2048)
    at src/engine/enginedeck.cpp:112
#10 0x00000000007e669d in EngineMaster::processChannels(int) (this=this@entry=0x1633fc0, iBufferSize=iBufferSize@entry=2048)
    at src/engine/enginemaster.cpp:318
#11 0x00000000007eee9b in EngineMaster::process(int) (this=0x1633fc0, iBufferSize=iBufferSize@entry=2048)
    at src/engine/enginemaster.cpp:350
#12 0x0000000000b6137b in SoundManager::onDeviceOutputCallback(unsigned int) (this=<optimized out>, iFramesPerBuffer=iFramesPerBuffer@entry=1024) at src/soundio/soundmanager.cpp:576
#13 0x0000000000b5d954 in SoundDevicePortAudio::callbackProcessClkRef(unsigned int, float*, float const*, PaStreamCallbackTimeInfo const*, unsigned long) (this=0x1d39070, framesPerBuffer=1024, out=0x42d83a00, in=<optimized out>, timeInfo=<optimized out>, statusFlags=<optimized out>) at src/soundio/sounddeviceportaudio.cpp:947
#14 0x00007ffff7872906 in NonAdaptingProcess (bp=bp@entry=0x42d80990, streamCallbackResult=streamCallbackResult@entry=0x7ffec37fdd28, hostInputChannels=0x0, hostOutputChannels=0x42d4ae90, framesToProcess=<optimized out>) at src/common/pa_process.c:871
#15 0x00007ffff7874914 in PaUtil_EndBufferProcessing (bp=bp@entry=0x42d80990, streamCallbackResult=streamCallbackResult@entry=0x7ffec37fdd28) at src/common/pa_process.c:1583
#16 0x00007ffff787d705 in CallbackThreadFunc (userData=0x42d80920) at src/hostapi/alsa/pa_linux_alsa.c:4328
#17 0x00007ffff23a75ca in start_thread () at /usr/lib64/libpthread.so.0
#18 0x00007fffefd470ed in clone () at /usr/lib64/libc.so.6
```

Be-ing commented on src/effects/native/reverbeffect.cpp:
```diff
@@ -74,48 +97,21 @@ void ReverbEffect::processChannel(const ChannelHandle& handle,
-                pOutput, pOutput, pState->crossfade_buffer, numSamples);
+    if (pState->sampleRate != sampleRate) {
+        // update the sample rate if it's changed
+        pState->reverb.init(sampleRate);
```
Won't this skip calling init() if the sample rate is 44.1 kHz (together with `float sampleRate  = 44100;` in the header file).

timrae replied:
oops yes thanks

timrae replied:
try again now

Be-ing replied:
This seems to be responsible for the segfault. Setting Mixxx to 96 kHz it does not segfault.

Be-ing commented on src/effects/native/reverbeffect.h:
```diff
@@ -17,23 +17,8 @@
-        SampleUtil::free(crossfade_buffer);
-    }
-
+    float sampleRate  = 0;
```
Does this need to be initialized to 0 here? Won't just `float sampleRate;` do? Setting a default of 0 kind of implies that a value of 0 is a valid value for this variable to have.

timrae replied:
it's usually considered bad practice to not initialize all variables in the constructor

timrae replied:
http://stackoverflow.com/questions/1069621/are-members-of-a-c-struct-initialized-to-0-by-default

timrae replied:
According to that link I think should also be value-initializing the other member (the effect). I'll amend the commit to put `MixxxPlateX2 reverb{};` instead of `MixxxPlateX2 reverb;`.

Be-ing commented:
Looks, and more importantly, sounds good to me! Thank you!

Chaining this after echo sounds nice. :)

Be-ing commented:
> I'm not sure, I saw that but it didn't look trivial so I skipped it. Feel free to make a new PR if this is something you're interested in working on ;)

It doesn't look too complicated, so I might try that. But I don't hear any issue with how it is.

rryan commented:
> Looks, and more importantly, sounds good to me! Thank you!

LGTMe-too, thanks! I'll merge once CI is happy.

timrae committed:
Fix the reverb effect
The previous implementation of the reverb effect was suffering from serious artifacts in terms of feedback / distortion that the original LADSPA plugin doesn't suffer from:
https://bugs.launchpad.net/mixxx/+bug/1647568

I updated the code base to use the latest version (0.9.24) from the original author (http://quitte.de/dsp/caps.html), and in doing this I disabled the unused code with preprocessor directives instead of deleting it, in order to make it easier to update in the future.

Next I tried to track down where the bug was. It turns out that the `SampleUtil::linearCrossfadeBuffers()` call was causing all the feedback problems, and it seems to be safe to remove this. In the future we should use ramping on the parameters.

This commit also adds the missing two controls for damping and blend, and remaps the meta knob from bandwidth to blend, which fixes the following bug:
https://bugs.launchpad.net/mixxx/+bug/1647775

Note that the meaning of "blend" has changed compared to the original source code, as we use it as a "send amount" instead of wet/dry. This has the advantage that the reverb tail is not suddenly cut off when the blend knob is turned down to zero.

In the PR (#1113) it was discussed that in the future we should probably implement wet/dry and send functions at the system level rather than forcing individual effects to implement it.

ronso0 commented:
Please make 'Send' one of the the first 3 parameters! With a regular 3-FX-knob controller, now it is impossible to effectively activate Reverb.
Parallel to Echo, Send should be first (most important) parameter.

Be-ing commented:
With the typical 3 effect knob controller, the send parameter is meant to be controlled with the effect's metaknob. I have actually proposed to make echo the same way as this in PR #1118. If you map your Reloop Terminal Mix 2 using [the EffectUnit JS object in the library](https://github.com/Be-ing/mixxx/blob/2e22cc52e05ed66a30ac7526484e278768f17a3d/res/controllers/Hercules-P32-scripts.js#L857) the arrangement of the parameters will make sense.

ronso0 commented:
Hm, yes, I'm aware of the new JS object but didn't implement it, yet.
I'll comment in #1118

ronso0 commented:
Actually, I see the advantages are dominating. What was bothering me is that I'd have to dump my current FX mapping (working well for parameters 1-3) and that now, one would have to switch between MetaKnob/single parameter mode to adjust Send and other parameters.
But never mind, I'll get used to that. Thanks for the effort! Both MetaKnob and the JS object will be very useful.
